src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.BuiltinOperatorSymbolKey.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial struct SymbolKey
+    {
+        private sealed class BuiltinOperatorSymbolKey : AbstractSymbolKey<IMethodSymbol>
+        {
+            public static readonly BuiltinOperatorSymbolKey Instance = new();
+
+            public sealed override void Create(IMethodSymbol symbol, SymbolKeyWriter visitor)
+            {
+                visitor.WriteString(symbol.Name);
+                visitor.WriteSymbolKey(symbol.ReturnType);
+                visitor.WriteParameterTypesArray(symbol.Parameters);
+            }
+
+            protected sealed override SymbolKeyResolution Resolve(
+                SymbolKeyReader reader, IMethodSymbol? contextualSymbol, out string? failureReason)
+            {
+                var name = reader.ReadRequiredString();
+
+                var returnType = reader.ReadSymbolKey(contextualSymbol?.ReturnType, out var returnTypeFailureReason);
+                using var parameterTypes = reader.ReadSymbolKeyArray<IMethodSymbol, ITypeSymbol>(
+                    contextualSymbol,
+                    static (contextualSymbol, i) => SafeGet(contextualSymbol.Parameters, i)?.Type,
+                    out var parameterTypesFailureReason);
+
+                if (returnTypeFailureReason != null)
```

Comment:
Can we check this before reading parameterTypes?

Reply:
You have to read what you wrote, regardless of whether you could do anything with it, or the position of the stream will be wrong.

Reply:
great question!  the answer is 'no'. You can think of the 'reading' as moving a pointer through the stream.  we have to always read to the end (even on failure).  this is because higher parts may re-read pieces (As they're resolving against different symbols).  It's definitely subtle.